skp_generator/build.sh

```diff
@@ -0,0 +1,66 @@
+# Copyright 2014 The Flutter Authors. All rights reserved.
+# Use of this source code is governed by a BSD-style license that can be
+# found in the LICENSE file.
+
+# Output each command, and quit immediately if any fail.
+set -ex
+
+# This function deletes everything except the skps directory and
+# what's already checked into git. This includes, for example, the
+# output of `flutter create`, any locally-checked-out flutter in this
+# directory, and the `build` directory (all of these get created by
+# subsequent action in this script).
+#
+# This function is called immediately on startup below, as well as
+# automatically when the script exits (see the "trap" below).
+function cleanup {
+    git clean -xffdq -e skps
+}
+
+cleanup
+#trap cleanup EXIT # COMMENT OUT THIS LINE TO LEAVE DEVELOPMENT ARTIFACTS (see README)
+
+# If Flutter isn't locally installed, then obtain it and add it to the path.
+if ! command -v flutter
+then
+    pushd ..
+    git clone https://github.com/flutter/flutter.git
+    export PATH="$PWD/flutter/bin:$PATH"
+    popd
+fi
+
+# Configure the application to be ready to run.
+flutter create .
+flutter pub add vm_service
+
+# Remove any previous output and prepare the directory.
+rm -rf test skps
+mkdir skps
+
+# Check for analysis errors. This script fails if this finds any errors.
+flutter analyze
+
+# Prepare the application for execution.
+flutter build bundle
+
+# This determines the type of binary to use. We try to run the
+# application on the local host but using the headless binary so that
+# there's no need to have an SDK installed, a device, a graphics
+# environment, etc. In particular this means it can run in CI.
+case "$(uname -s)" in
+    Darwin)
+        PLATFORM=darwin-x64
+        ;;
+    Linux)
+        PLATFORM=linux-x64
+        ;;
+    MINGW*)
+        PLATFORM=windows-x64
+        ;;
+esac
+
+# Find the flutter_tester binary and runs it.
+# The output will be placed in the skps/ directory.
+# We use --run-forever because the app terminates itself when it's done.
+# We use --non-interactive for obvious reasons.
+$(dirname $(which flutter))/cache/artifacts/engine/$PLATFORM/flutter_tester --run-forever --non-interactive --packages=.dart_tool/package_config.json --flutter-assets-dir=build/flutter_assets/ build/flutter_assets/kernel_blob.bin
```

Comment:
I think this might fail if you're using a symlink. @gspencergoog ?

Reply:
Yeah, it probably would.  You'll probably want to use `type -P flutter` instead of `which` (since `which` isn't the same everywhere), and then do something like this to find the actual directory where it exists.  I know this seems overly complicated, but we've looked at a lot of options, and this worked most reliably on the most systems for finding the location of the actual file so you could take its dirname safely.

```bash
# Needed because if it is set (e.g. incorrectly exported), cd may print the
# path it changed to.
unset CDPATH

# On Mac OS, readlink -f doesn't work, so follow_links traverses the path one
# link at a time, and then cds into the link destination and find out where it
# ends up.
#
# The returned filesystem path must be a format usable by Dart's URI parser,
# since the Dart command line tool treats its argument as a file URI, not a
# filename. For instance, multiple consecutive slashes should be reduced to a
# single slash, since double-slashes indicate a URI "authority", and these are
# supposed to be filenames. There is an edge case where this will return
# multiple slashes: when the input resolves to the root directory. However, if
# that were the case, we wouldn't be running this shell, so we don't do anything
# about it.
#
# The function is enclosed in a subshell to avoid changing the working directory
# of the caller.
function follow_links() (
  cd -P "$(dirname -- "$1")"
  file="$PWD/$(basename -- "$1")"
  while [[ -h "$file" ]]; do
    cd -P "$(dirname -- "$file")"
    file="$(readlink -- "$file")"
    cd -P "$(dirname -- "$file")"
    file="$PWD/$(basename -- "$file")"
  done
  echo "$file"
)

# Absolute path to actual flutter location.
FLUTTER="$(follow_links "$(type -P flutter)")"
BIN_DIR="$(cd "${FLUTTER%/*}" ; pwd -P)"
```

Reply:
I'm not sure what problem we're trying to solve here. The use case for this script is Flutter team members who work on Flutter, and CI. In CI the Flutter tool is downloaded by this script itself, no symlinks involved. For Flutter team members, the workflow is to have the flutter shell script on your path, no symlinks involved.

Reply:
I've integrated the code above into the script.

Reply:
FWIW, I have had setups for flutter in the past that had a link to the `bin/flutter` script from a `cmd` directory that was in my PATH.  I don't have that setup anymore, but I don't think it's unreasonable.

Reply:
Yeah, we have this kind of logic in most of the Flutter shell scripts so that people who symlink their flutter SDK to more rapidly switch between different versions (without downloading all the artifacts for each switch) don't get broken.

Reply:
k